cmd/spire-agent/cli/run/run_posix_test.go

```diff
@@ -5,15 +5,121 @@ package run
 
 import (
 	"bytes"
+	"fmt"
+	"io"
 	"os"
+	"syscall"
 	"testing"
 
 	"github.com/hashicorp/hcl/hcl/printer"
 	"github.com/spiffe/spire/pkg/agent"
+	commoncli "github.com/spiffe/spire/pkg/common/cli"
+	"github.com/spiffe/spire/pkg/common/log"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 )
 
+func TestCommand_Run(t *testing.T) {
+	testTempDir := t.TempDir()
+	testDataDir := fmt.Sprintf("%s/data", testTempDir)
+	testAgentSocketDir := fmt.Sprintf("%s/spire-agent", testTempDir)
+
+	type fields struct {
+		logOptions         []log.Option
+		env                *commoncli.Env
+		allowUnknownConfig bool
+	}
+	type args struct {
+		args []string
+	}
+	type want struct {
+		code               int
+		dataDirCreated     bool
+		agentUdsDirCreated bool
+	}
+	tests := []struct {
+		name     string
+		osTarget string
+		fields   fields
+		args     args
+		want     want
+	}{
+		{
+			name:     "don't create any dir when error loading config",
+			osTarget: "linux",
+			args: args{
+				args: []string{},
+			},
+			fields: fields{
+				logOptions: []log.Option{},
+				env: &commoncli.Env{
+					Stderr: io.Discard,
+				},
+				allowUnknownConfig: false,
+			},
+			want: want{
+				code:               1,
+				agentUdsDirCreated: false,
+				dataDirCreated:     false,
+			},
+		},
+		{
+			name:     "creates spire-agent uds and data dirs",
+			osTarget: "linux",
+			args: args{
+				args: []string{
+					"-config", "../../../../test/fixture/config/agent_run_posix.conf",
+					"-trustBundle", "../../../../conf/agent/dummy_root_ca.crt",
+					"-dataDir", testDataDir,
+					"-socketPath", fmt.Sprintf("%s/spire-agent/api.sock", testTempDir),
+				},
+			},
+			fields: fields{
+				logOptions: []log.Option{},
+				env: &commoncli.Env{
+					Stderr: io.Discard,
+				},
+				allowUnknownConfig: false,
+			},
+			want: want{
+				code:               1,
+				agentUdsDirCreated: true,
+				dataDirCreated:     true,
+			},
+		},
+	}
+	for _, testCase := range tests {
+		t.Run(testCase.name, func(t *testing.T) {
+			os.RemoveAll(testDataDir)
+			os.RemoveAll(testAgentSocketDir)
+
+			cmd := &Command{
+				logOptions:         testCase.fields.logOptions,
+				env:                testCase.fields.env,
+				allowUnknownConfig: testCase.fields.allowUnknownConfig,
+			}
+
+			result := cmd.Run(testCase.args.args)
+			assert.Equal(t, testCase.want.code, result)
+
+			_, agentSocketDirErr := os.Stat(testAgentSocketDir)
+
+			if testCase.want.agentUdsDirCreated {
+				assert.Nilf(t, agentSocketDirErr, "spire-agent uds dir should be created")
```

Comment:
When asserting errors, the specific functions should be used to have a better assertion and diagnostics. Error and NoError functions are provided for this purpose.

```suggestion
				assert.NoError(t, agentSocketDirErr, "spire-agent uds dir should be created")
```